File: Data_Structures/Linked_List/Singly_Linked_List/PalindromeLinkedList.js
 //classes to create the linklist
 class LinkedListNode {
    constructor(char, next) {
      this.value = char;
      this.next = next;
    }
  }
  class LinkedList {
    constructor(chars) {
      const len = chars.length;
      let currentNode = null;
      for (let i = len - 1; i >= 0; i--) {
        let node = new LinkedListNode(chars[i], currentNode);
        currentNode = node;
      }
      this.head = currentNode;
    }
  //function to check if the linked list iws a palindrome
    isPalindrome() {
      let center, slow, quick;
      center = slow = quick = this.head;
      if (slow.next == null) {
        return true;
      }
      while(quick.next != null) {
        slow = slow.next;
        slow.prev = center;
        center = slow
        quick = quick.next
        if (quick.next == null) {
          // even number
          center = slow.prev;
        } else {
          quick = quick.next;
        }
      }
      do {
        if(slow.value !== center.value) {
          return false;
        }
        slow = slow.next;
        center = center.prev;
      } while(slow != null) 
      return true;
    }
  }
var list[];
for(var i =0;i<1)
{
var x=window.prompt("enter the element: ",if done enter DONE);
if(x!=='DONE')
{
 list.push(x);
}
 else{
  i++;
 }
}
  console.log(new LinkedList(list).isPalindrome());

File: Data_Structures/Linked_List/Singly_Linked_List/Java/PalindromeLinkedList.java


  class ListNode {
     int val;
     ListNode next;
     ListNode() {}
     ListNode(int val) { this.val = val; }
     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 }

class LinkedList {
	private ListNode head;
	public void addToTheLast(ListNode node) {
		 
		if (head == null) {
			head = node;
		} else {
			ListNode temp = head;
			while (temp.next != null)
				temp = temp.next;
 
			temp.next = node;
		}
	}
 
	
    public boolean isPalindrome(ListNode head) {
    ListNode fast = head, slow = head;
    while (fast != null && fast.next != null) {
        fast = fast.next.next;
        slow = slow.next;
    }
    if (fast != null) { // odd nodes: let right half smaller
        slow = slow.next;
    }
    slow = reverse(slow);
    fast = head;
    
    while (slow != null) {
        if (fast.val != slow.val) {
            return false;
        }
        fast = fast.next;
        slow = slow.next;
    }
    return true;
}

public ListNode reverse(ListNode head) {
    ListNode prev = null;
    while (head != null) {
        ListNode next = head.next;
        head.next = prev;
        prev = head;
        head = next;
    }
    return prev;
}
}


 class PalindromLinkedList{
 public static void main(String args[]){
	 LinkedList list = new LinkedList();
	 ListNode head=new ListNode(1);
		list.addToTheLast(head);
		list.addToTheLast(new ListNode(2));
		list.addToTheLast(new ListNode(1));
		list.addToTheLast(new ListNode(4));
		list.addToTheLast(new ListNode(1));

  
  if(list.isPalindrome(head)){
    System.out.println("Linked List is Palindrome!");
 }else{
    System.out.println("Linked List is Not Palindrome!");
  }
}
 
}
